python/sglang/srt/utils/hf_transformers_utils.py

```diff
@@ -392,12 +401,148 @@ def get_context_length(config):
 _FAST_LLAMA_TOKENIZER = "hf-internal-testing/llama-tokenizer"
 
 
+def find_local_tokenizer_snapshot_dir(
+    model_name_or_path: str,
+    cache_dir: Optional[str],
+    allow_patterns: List[str],
+    revision: Optional[str] = None,
+) -> Optional[str]:
+    """If the tokenizer files are already local, skip downloading and return the path.
+    Only applied in CI.
+    """
+    if not is_in_ci():
+        return None
+
+    if os.path.isdir(model_name_or_path):
+        if is_in_ci():
+            print(
+                f"[TOKENIZER CACHE] {model_name_or_path} is already a local directory, skipping cache check"
+            )
+        return None
+
+    print(f"[TOKENIZER CACHE] Checking for cached tokenizer: {model_name_or_path}")
+    found_local_snapshot_dir = None
+
+    # Check custom cache_dir (if provided)
+    if cache_dir:
+        try:
+            repo_folder = os.path.join(
+                cache_dir,
+                huggingface_hub.constants.REPO_ID_SEPARATOR.join(
+                    ["models", *model_name_or_path.split("/")]
+                ),
+            )
+            rev_to_use = revision
+            if not rev_to_use:
+                ref_main = os.path.join(repo_folder, "refs", "main")
+                if os.path.isfile(ref_main):
+                    with open(ref_main) as f:
+                        rev_to_use = f.read().strip()
+            if rev_to_use:
+                rev_dir = os.path.join(repo_folder, "snapshots", rev_to_use)
+                if os.path.isdir(rev_dir):
+                    found_local_snapshot_dir = rev_dir
+        except Exception as e:
+            logger.warning(
+                "Failed to find local snapshot in custom cache_dir %s: %s",
+                cache_dir,
+                e,
+            )
+
+    # Check default HF cache as well
+    if not found_local_snapshot_dir:
+        try:
+            rev_dir = find_local_repo_dir(model_name_or_path, revision)
+            if rev_dir and os.path.isdir(rev_dir):
+                found_local_snapshot_dir = rev_dir
+        except Exception as e:
+            logger.warning("Failed to find local snapshot in default HF cache: %s", e)
+
+    # If local snapshot exists, validate it contains at least one tokenizer file
+    # matching allow_patterns before skipping download.
+    if found_local_snapshot_dir is None:
+        return None
+
+    local_tokenizer_files: List[str] = []
+    try:
+        for pattern in allow_patterns:
+            local_tokenizer_files.extend(
+                glob.glob(os.path.join(found_local_snapshot_dir, pattern))
+            )
+    except Exception as e:
+        logger.warning(
+            "Failed to scan local snapshot %s with patterns %s: %s",
+            found_local_snapshot_dir,
+            allow_patterns,
+            e,
+        )
+        local_tokenizer_files = []
+
+    if len(local_tokenizer_files) > 0:
```

Comment:
We need to make this condition much more restrictive. Otherwise, it will try to load corrupted weighs.
see also https://github.com/sgl-project/sglang/pull/11248

Some ideas: check hash, check incomplete files, etc
https://github.com/sgl-project/sglang/blob/0f8bd55f3e8ffebe30f3ebbaf4e1bc5379def69e/python/sglang/srt/model_loader/weight_validation.py#L13
https://github.com/sgl-project/sglang/blob/0f8bd55f3e8ffebe30f3ebbaf4e1bc5379def69e/scripts/ci/cleanup_hf_cache.py#L49-L55